Bundle updates in MoM before sending to server and vice-versa

hirenvadalia opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
* Currently MoM send each message separately to server so for example in case of small jobs will have lots of individual messages to process in server
* Currently MoM send session id (aka first resc used information) as soon as job starts, but if job is very short then there will be obit ready immediately followed by session id information to server, so we can club both information in one message - obit message
* Now if there are multiple resc used or resc used from hook message for same job in one mom main cycle, currently MoM send all to server and server has to process each of them, but we can merge all into one and send only one update to server - less messages + less to process by server
* Currently if multiple obit comes then server replies to them one by one, instead we can club them into only 2 messages, 1 for ack obits and 2nd for reject/bad obits (if any)
* Currently MoM spends its most time in mom_set_use_all() as it sets resc usage on each job known to mom, but its not needed if job is just queued to mom or in stage-in process or ended and waiting for server's reply or in stage-out process, we need to set resc usage only on actually running jobs
* Currently on obit server sends obit ack, stageout req (if any), stage del req (if any) and delete job req to MoM, but if there is nothing to stage out/del and there is no end hook to run then we can expedite end job process on both side as both server/MoM has all needed information to find whether anything to stage out/del or end hook to run or not. This will reduce number messages + fast end job on both side without waiting for each other. (2nd commit - "Express end job")

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
* Bundling + express end job:
  * if there is nothing to stage out/del and no end hook to run purge job once server acks obit in mom and in do end job processing in server
  * don't send any obit or resc used immediately, instead put it in pending list (aka enqueue for send)
  * Now in main loop bundle similar updates in groups (like obit, resc used or resc used from hook) from pending list
  * during bundling, for resc used and resc used from hook, check for time of creation of usage, if it is less than 2s then don't include in bundle. Means delay sending resc used and resc used from hook for 2s but don't delay obit
  * once bundle is ready, send it (so at-most will have only 3 messages to send to server (in one mom main cycle) with multiple jobs information in it instead each individual message)
  * Now for jobs which ends in <= 2s there will be only 2 messages (ack to queuejob and obit) to server instead 4 (ack to queuejob, session id, obit and ack to del job)
    * On server side 2 less message to process (session id and ack to deljob), one less message to send (del job)
* call mom_set_use() for only actually running jobs in mom_set_use_all
* other few fixes:
  * removed of trk_discarding from tracking table as its not needed due to trk_psubjob->ji_discarding
  * small optimization in vn_merge2(), mom_vnlp_report() and mom_CPUs_report()
  * while developing this changes, I found two test failures due to race condition, fixed them also

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
* None

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, CENTOS8
Platforms: UBUNTU1804,SUSE15,CENTOS8
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|3414|1815|5|0|0|1|1809|

(NOTE: 5 failures are known issues)
* Valgrind results - [valgrind.txt](https://github.com/openpbs/openpbs/files/5112443/valgrind.txt)

* Perf results
  * 20 moms each 100 ncpus
  * sched on + qsub -koe -- /bin/date job
  * parallel jobs submission with 20 threads 5k/thread evenly distributed among 5 users 
     * script [submit-jobs.txt](https://github.com/openpbs/openpbs/files/5141555/submit-jobs.txt) (called with args `20 5000 15001`)


||Try1|Try2|Try3|Avg.|
|-|-|-|-|-|
|Before changes|duration: 0:04:08<br>end_rate: 403.23/s<br>run_rate: 403.23/s<br>submit_rate: 1754.39/s<br>throughput: 403.23/s|duration: 0:04:04<br>end_rate: 409.84/s<br>run_rate: 409.84/s<br>submit_rate: 1754.39/s<br>throughput: 409.84/s|duration: 0:04:12<br>end_rate: 396.83/s<br>run_rate: 396.83/s<br>submit_rate: 1818.18/s<br>throughput: 396.83/s|duration: 0:04:08<br>end_rate: 403.30/s<br>run_rate: 403.30/s<br>submit_rate: 1775.65/s<br>throughput: 403.30/s|
|After changes|duration: 0:02:59<br>end_rate: 558.66/s<br>run_rate: 558.66/s<br>submit_rate: 1960.78/s<br>throughput: 558.66/s|duration: 0:02:58<br>end_rate: 561.80/s<br>run_rate: 561.80/s<br>submit_rate: 1960.78/s<br>throughput: 561.80/s|duration: 0:02:57<br>end_rate: 564.97/s<br>run_rate: 564.97/s<br>submit_rate: 1960.78/s<br>throughput: 564.97/s|duration: 0:02:57<br>end_rate: 561.81/s<br>run_rate: 561.81/s<br>submit_rate: 1960.78/s<br>throughput: 561.81/s|

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+	}
+}
+
+static int
```
add doxygen

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+
+
+				/* The following 2 temp variables will be set to 1
+				 * if there's an error accuumlating resources_used
```
existing code - but let us remove the extra 'u'

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+{
+	ruu *cur;
+	ruu *next;
+#define RESCUSED_SEND_DELAY 2
```
why don't we use a variable here?

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+	*prused = NULL;
+	*rh_cnt = 0;
+	*prhused = NULL;
+	*o_cnt = 0;
```
would saying obit_count be more readable?

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+
+/**
+ * @brief
+ * 	Send the amount of resouces used by jobs to the server
```
resources

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+			if (ret != DIS_SUCCESS)
+				goto err;
+		}
+		ret =diswsi(server_stream, rud->ru_status);
```
space after '='

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+static void
+send_pending_updates(void)
+{
+	send_pending_ruu();
```
why don't we call send_pending_ruu() directly from the caller?

prakashcv13 commented on src/resmom/mom_server.c:
```diff
@@ -885,45 +882,83 @@ is_request(int stream, int version)
-			if (pjob && !pjob->ji_hook_running_bg_on && (pjob->ji_qs.ji_substate != JOB_SUBSTATE_PRERUN)) {
-				log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_NOTICE, jobid, "Job removed, Server rejected Obit");
-				mom_deljob(pjob);
+			while(njobs--) {
```
space after while

prakashcv13 commented on src/resmom/mom_server.c:
```diff
@@ -885,45 +882,83 @@ is_request(int stream, int version)
+			DBPRT(("%s: IS_ACKOBIT njobs: %d\n", __func__, njobs))
+			log_eventf(PBSEVENT_DEBUG3, PBS_EVENTCLASS_JOB, LOG_DEBUG, __func__, "received ack obits = %d", njobs);
+
+			while(njobs--) {
```
space after while

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1603,7 +1604,7 @@ def test_ASAP_resv_request_same_time(self):
                                             jid=jid2)
         rid1_q = rid1.split('.')[0]
-        exp_attr = {'reserve_state': (MATCH_RE, "RESV_CONFIRMED|2"),
+        exp_attr = {'reserve_state': (MATCH_RE, "RESV_RUNNING|5"),
```
didn't understand this change

hirenvadalia replied:
Actually this is test fix failure i encountered while I was working on these changes, but I will double check whether this change is needed or not...

hirenvadalia replied:
I re-tested this, and with `RESV_CONFIRMED|2`, test some time passes (6/10 times) and some time fails (4/10 times) but with `RESV_RUNNING|5` test always passes. So I think we need this fix.

hirenvadalia replied:
as per offline discussion, I have added proper fix for race condition in test case

prakashcv13 commented on src/server/job_func.c:
```diff
@@ -1957,3 +1963,428 @@ find_aoe_from_request(resc_resv *presv)
+			break;
+		*(to+i) = *(from+i);
+	}
+	*(to+i) = '\0';
```
spaces around all operators

prakashcv13 commented on src/server/node_manager.c:
```diff
@@ -2134,95 +2128,122 @@ stat_update(int stream)
+			int is_reject = 0;
+
+			DBPRT(("recv_job_obit: decoded obit for %s\n", rused.ru_pjobid))
+			job_obit(&rused, stream, &is_reject);
```
why use an in-out parameter? shall we have job_obit return an int.

prakashcv13 commented on src/server/node_manager.c:
```diff
@@ -2134,95 +2128,122 @@ stat_update(int stream)
+			free(reject_list[i]);
+			reject_list[i] = NULL;
 		}
+		dis_flush(stream);
```
instead of sending two different replies, can we send only one?
the reply will be job_id followed by 0 or 1 indicating if its is a bad obit or not.
in the MoM, we introduce a new message type like IS_OBITREPLY

prakashcv13 commented on src/server/node_manager.c:
```diff
@@ -6843,19 +6867,19 @@ remove_job_index_from_mom(job *pjob, struct pbsnode *pnode)
+			continue;
+		psvrmom = (mom_svrinfo_t *)(pnode->nd_moms[i]->mi_data);
+
+		for (j=0; j<psvrmom->msr_jbinxsz; j++) {
```
spaces around '='

prakashcv13 reviewed (changes requested):
just have one suggestion
rest are minor and cosmetic comments

hirenvadalia commented:
@prakashcv13 Thanks for through review, all comments accepted, I will address them soon...

hirenvadalia commented on src/resmom/catch_child.c:
```diff
@@ -1242,66 +410,26 @@ send_obit(job *pjob, int exval)
 	} else {
-		DBPRT(("send_obit: job %s runcount %ld exval %d\n", pjob->ji_qs.ji_jobid, pjob->ji_wattr[(int)JOB_ATR_runcount].at_val.at_long, exval))
+		DBPRT(("send_obit: job %s runcount %ld exval %d\n", pjob->ji_qs.ji_jobid, pjob->ji_wattr[(int) JOB_ATR_runcount].at_val.at_long, exval))
 	}
```
we need those braces since those block has only DBPRTs and they are not defined in release mode so it will give errors on compile...

hirenvadalia commented on src/resmom/mom_main.c:
```diff
@@ -6437,6 +6445,919 @@ kill_job(job *pjob, int sig)
+	if (pjob->ji_wattr[JOB_ATR_Comment].at_flags & ATR_VFLAG_SET) {
+		prused->ru_comment = strdup(pjob->ji_wattr[JOB_ATR_Comment].at_val.at_str);
+		if (prused->ru_comment == NULL) {
+			FREE_RUU(prused);
```
@prakashcv13 assuming you are return NULL if we fails to encode comment in ruu, then we are already doing that. If you are asking not to return NULL and continue without comment in ruu, then I am not sure about that, need to check on windows as sending comment is only for windows not linux.

prakashcv13 replied:
i was just thinking that setting of comment is not something that should stop us from continuing. If it makes more sense to not continue, please feel free to ignore my comment.

hirenvadalia replied:
Thanks for confirming @prakashcv13, I will double check on windows that without comment is not giving any problem, and will change accordingly

hirenvadalia replied:
So I checked this, and its turns out that comment in updates is not used at all in stat_update but only used while job_obit and that only for one case, which is if job fails due to bad password and on server side job goes in hold state, so I think we can send update if strdup() fails as it will get included in next updates (like other attrs)...

agrawalravi90 commented:
@hirenvadalia Can you please post some performance numbers?

hirenvadalia commented:
> @hirenvadalia Can you please post some performance numbers?

Sure @agrawalravi90, I am just waiting for some time + our in-house perf machine to become free.

hirenvadalia commented:
@prakashcv13 I have addressed your comments, except two - 1st for continue on job comment encode failure and 2nd - test fix, I will get back to you on that soon. But meanwhile you can proceed with review for other comments.

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -2655,30 +2649,21 @@ do_mom_action_script(int	ae,	/* index into action table */
-			ma->ma_name, ma->ma_script);
-		log_event(PBSEVENT_ADMIN, PBS_EVENTCLASS_JOB, LOG_INFO,
-			pjob->ji_qs.ji_jobid, log_buffer);
+		log_eventf(PBSEVENT_ADMIN, PBS_EVENTCLASS_JOB, LOG_INFO, pjob->ji_qs.ji_jobid, "action %s script %s does not exist", ma->ma_name, ma->ma_script);
```
let's break this down like the one below.

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -3041,40 +2972,32 @@ do_mom_action_script(int	ae,	/* index into action table */
-			ptask->ti_qs.ti_task);
-		log_event(PBSEVENT_ADMIN, PBS_EVENTCLASS_JOB, LOG_INFO,
-			pjob->ji_qs.ji_jobid, log_buffer);
+	} else if (ae == RestartAction) {
```
braces not needed.

prakashcv13 commented on src/resmom/mom_main.c:
```diff
@@ -7044,6 +6897,18 @@ encode_used(job *pjob, pbs_list_head *phead)
+ *
+ */
 static int
 is_in_old_ruu(ruu *old, char *name)
```
as discussed offline, let us remove this function and use the logic in the caller (just one).

hirenvadalia commented:
Hey @agrawalravi90 @prakashcv13, I have added performance results in PR description, also updated latest regression results in same. Please have a look and let me know if you have any other tests in mind.

subhasisb commented on src/resmom/mom_hook_func.c:
```diff
@@ -571,7 +571,7 @@ copy_file_and_set_owner(char *src_file, char *dest_file, job *pjob) {
 	READS_MASK|WRITES_MASK|STANDARD_RIGHTS_REQUIRED) \
 							      == 0 ) {
-		snprintf(log_buffer, LOG_BUF_SIZE, "Unable to change permissions of the file for user: %s, file: %s", 
+		snprintf(log_buffer, LOG_BUF_SIZE, "Unable to change permissions of the file for user: %s, file: %s",
```
use log_eventf

subhasisb commented on src/resmom/mom_hook_func.c:
```diff
@@ -1059,7 +1059,7 @@ run_hook(hook *phook, unsigned int event_type, mom_hook_input_t *hook_input,
 		READS_MASK|WRITES_MASK|STANDARD_RIGHTS_REQUIRED) \
 								      == 0 ) {
-			snprintf(log_buffer, LOG_BUF_SIZE, "Unable to change permissions of the script file for user: %s, file: %s", 
+			snprintf(log_buffer, LOG_BUF_SIZE, "Unable to change permissions of the script file for user: %s, file: %s",
```
can use log_eventf in these cases?

hirenvadalia commented on src/include/net_connect.h:
```diff
@@ -96,30 +96,28 @@ typedef unsigned long pbs_net_t;        /* for holding host addresses */
-
-#define IS_CMD          40
-#define IS_CMD_REPLY    41
+#define IS_CMD                          1
```
For this changes no, but when we did mom reversal changes, we changed numbers to make it sequential and removed few unused for long time, but we missed to change for IS_CMD and IS_CMD_REPLY, so I changed now

subhasisb replied:
ok, perhaps we need to move to flat buffers soon to ensure that these headers are all backward compatible and versioned, so that we can make rolling upgrades are reality without too much effort

agrawalravi90 reviewed (dismissed):
Minor comments.

Also, just a suggestion for future, you have done a lot of refactoring and moved functions around, which is great, but it makes it really hard for reviewers to know whether the code change that I'm looking at is new code or just code that was moved/refactored, if it's just code which was moved, I wouldn't bother reviewing it. So, it might help breaking out refactoring into a separate commit.

subhasisb commented on src/include/net_connect.h:
```diff
@@ -96,30 +96,28 @@ typedef unsigned long pbs_net_t;        /* for holding host addresses */
-
-#define IS_CMD          40
-#define IS_CMD_REPLY    41
+#define IS_CMD                          1
```
actually enum could be dangerous if you suddenly comment one of the numbers out, the next in the series would have their values changed. I would say keeping the numbers the same by using hash defines and providing explicit values to them makes for more interoperability (not that we support that first class, but we would need to going forward)

bhroam replied:
The scheduler uses enums for things like this.  The thing is it will explicitly set the values.  If you were to delete one symbol, nothing changes.  It more or less is the same thing, except that when you are debugging you see the symbol instead of the number.

subhasisb replied:
yeah - that would work, but takes away most of the usefulness of having an enum. Yeah i see the nice thing in debugging

hirenvadalia replied:
After discussion with @subhasisb, right now we are not changing this to enum, but at time of flat-buffer (or in future) will change to enum if needed...

subhasisb commented:
> Minor comments.
> 
> Also, just a suggestion for future, you have done a lot of refactoring and moved functions around, which is great, but it makes it really hard for reviewers to know whether the code change that I'm looking at is new code or just code that was moved/refactored, if it's just code which was moved, I wouldn't bother reviewing it. So, it might help breaking out refactoring into a separate commit.

I agree, i think we should make the refactorings a separate PR or atleast separate commits, so that code review is easier. It is difficult to see code changes otherwise

hirenvadalia commented:
> > Minor comments.
> > Also, just a suggestion for future, you have done a lot of refactoring and moved functions around, which is great, but it makes it really hard for reviewers to know whether the code change that I'm looking at is new code or just code that was moved/refactored, if it's just code which was moved, I wouldn't bother reviewing it. So, it might help breaking out refactoring into a separate commit.
> 
> I agree, i think we should make the refactorings a separate PR or atleast separate commits, so that code review is easier. It is difficult to see code changes otherwise

Sorry for mess but noted, will keep this in mind from next time... I really wish code formatting project gets in soon :smile:

hirenvadalia commented:
##### Latest regression result:
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, CENTOS8
Platforms: UBUNTU1804,SUSE15,CENTOS8
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|3543|1726|3|0|0|0|1723|

(NOTE: 3 failures are known issue as below)

##### Regression results of master:
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, CENTOS8
Platforms: UBUNTU1804,SUSE15,CENTOS8
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|3542|1726|3|0|0|0|1723|

subhasisb commented:
I see u have addressed Ravi's feedback about moving out of the mom_main.c file as well. And we can take up the change to enum later. All tests have passed, merging.
